src/brevitas/core/function_wrapper/learned_round.py

```diff
@@ -0,0 +1,95 @@
+# Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""
+Different implementations for LearnedRound
+"""
+
+from typing import List, Optional
+
+import torch
+
+import brevitas
+from brevitas import config
+from brevitas.function.ops_ste import floor_ste
+
+
+class LearnedRoundHardSigmoid(brevitas.jit.ScriptModule):
+    """
+    HardSigmoid implementation for LearnedRound learned parameter
+    """
+    __constants__ = ['learnedround_zeta', 'learnedround_gamma']
+
+    def __init__(self, learnedround_zeta: float = 1.1, learnedround_gamma: float = -0.1) -> None:
+        super(LearnedRoundHardSigmoid, self).__init__()
+        self.learnedround_zeta = learnedround_zeta
+        self.learnedround_gamma = learnedround_gamma
+
+    @brevitas.jit.script_method
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        p = torch.sigmoid(x)
+        p = torch.clamp(
+            p * (self.learnedround_zeta - self.learnedround_gamma) + self.learnedround_gamma,
+            0.0,
+            1.0)
+        return p
+
+
+class LearnedRoundSigmoid(brevitas.jit.ScriptModule):
+    """
+    Sigmoid implementation for LearnedRound learned parameter. Supports for temperature factor
+    """
+    __constants__ = ['learnedround_temperature']
+
+    def __init__(self, learnedround_temperature: float = 1.) -> None:
+        super(LearnedRoundSigmoid, self).__init__()
+        assert learnedround_temperature != 0, 'Temperature should be different than 0'
+        self.learnedround_temperature = learnedround_temperature
+
+    @brevitas.jit.script_method
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        p = torch.sigmoid(x / self.learnedround_temperature)
+        return p
+
+
+class LearnedRoundSte(brevitas.jit.ScriptModule):
+    """
+    This Module implements LearnedRound representation, where each weight has a learnable parameter
+    that decides if "ceil" or "floor" rounding type has to be used.
+    """
+
+    def __init__(
+            self,
+            tracked_parameter_list: List[torch.nn.Module],
+            learnedround_impl: torch.nn.Module,
+            learnedround_init: Optional[torch.Tensor] = None) -> None:
+        super(LearnedRoundSte, self).__init__()
+        if len(tracked_parameter_list) > 1:
+            raise RuntimeError('LearnedRound does not support shared quantizers')
+        self.learnedround_impl = learnedround_impl
+        if learnedround_init == None:
+            self.value = torch.nn.Parameter(torch.full(tracked_parameter_list[0].shape, 0.))
```

Comment:
let's move this to a value function called learned_round_init where the enum is solved too that depends on tracked_parameter_list, that also does the check for shared quantizers. This way in the init here we just need to take in learned_round_init.